crates/turborepo-lib/src/run/watch.rs

```diff
@@ -140,23 +131,25 @@ impl WatchClient {
 
         let signal_subscriber = self.handler.subscribe().ok_or(Error::NoSignalHandler)?;
 
-        let (changed_pkgs_tx, mut changed_pkgs_rx) = watch::channel(ChangedPackages::default());
+        // We explicitly use a tokio::sync::Mutex here to avoid deadlocks.
+        // If we used a std::sync::Mutex, we could deadlock by spinning the lock
+        // and not yielding back to the tokio runtime.
+        let changed_packages = Mutex::new(RefCell::new(ChangedPackages::default()));
 
         let event_fut = async {
             while let Some(event) = events.next().await {
                 let event = event?;
-                Self::handle_change_event(&changed_pkgs_tx, event.event.unwrap()).await?;
+                Self::handle_change_event(&changed_packages, event.event.unwrap()).await?;
             }
 
             Err(Error::ConnectionClosed)
         };
 
         let run_fut = async {
             loop {
-                changed_pkgs_rx.changed().await?;
-                let changed_pkgs = { changed_pkgs_rx.borrow_and_update().clone() };
-
-                self.execute_run(changed_pkgs).await?;
+                if !changed_packages.lock().await.borrow().is_empty() {
```

Comment:
Does this spin? If `changed_packages` is in fact empty, won't this loop back around? It will yield the event loop, but the task will always be runnable as long as nothing else is holding the lock...

Reply:
If I understand correctly, if changed packages is empty, yes it’ll loop around, but on the next time it tries to lock, it’ll yield first to let the other task run and add to the changed packages. 

Could you clarify what you mean by the task being runnable?

Reply:
It will yield, but the tokio scheduler will always see this loop as ready to run, even if there's no work to be done. If instead we use a channel or some other notification mechanism, tokio can park this task until there is work to be done.